### PR TITLE
[ENH] [1/5] Header structure: replace specializations

### PR DESCRIPTION
**Pull request by ahendriksen**
This PR replaces the specialization headers with a split header structure. It is a more reviewable version of #1415.

This is achieved by following the strategy outlined in issue https://github.com/rapidsai/raft/issues/1416.

The following steps are performed:

**Preparatory**
- **Move** files around. 
  - [d9801e8](https://github.com/rapidsai/raft/pull/1437/commits/d9801e87f266554c2c561490faf0fe5d76e66165) Move header files with expensive function templates from `header.cuh` to `header-inl.cuh` to preserve history and minimize the diff in later comits.
  - [50b374d](https://github.com/rapidsai/raft/pull/1437/commits/50b374da5c579d7d160c15ebc0d2b1db105798eb) Move raft_runtime source files into separate directory. They would otherwise clash with source files added in later commits. 
- **Fix** existing issues in RAFT unearthed by this PR:
  - [8974ae3](https://github.com/rapidsai/raft/pull/1437/commits/8974ae32255e702de9d707da34b353b61bedbc23) Fix several locations where files were depending on transitive includes
  - [95ef31b](https://github.com/rapidsai/raft/pull/1437/commits/95ef31bc974b6b2b4739cedd6d4d287e9e0266be) Fix distance::getWorkspaceSize. It did not compile previously.
- **Prepare** for changes that are coming:
  - [7edcb6a](https://github.com/rapidsai/raft/pull/1437/commits/7edcb6a59da62216d9521e15c3848ba676588f31) Create a functor for `rbf_fin_op` that can be used to instantiate templates later on.
  - [71de7bd](https://github.com/rapidsai/raft/pull/1437/commits/71de7bd4e33e788c17bf7c27893d764e3c410b06) Similar move for distance functors used in `registers.cuh`.

**Meat of PR**: Use header structure as outlined in the [new documentation](https://github.com/ahendriksen/raft/blob/enh-inl-headers-slow/docs/source/developer_guide.md#header-organization-of-expensive-function-templates). 
- [Change RAFT_COMPILED from INTERFACE to PUBLIC](https://github.com/rapidsai/raft/pull/1437/commits/541cabc385d193c908bd2faf39e6a9ce3c58dfc1)
- [Define RAFT_EXPLICIT and RAFT_EXPLICIT_INSTANTIATE_ONLY](https://github.com/rapidsai/raft/pull/1437/commits/d81b14ee5e25ac78e042671323765659f5f693af)
- [Update docs](https://github.com/rapidsai/raft/pull/1437/commits/48ea76957413b9f3238fbac2ce456fde893c60e4)
- [Replace specializations by split headers](https://github.com/rapidsai/raft/pull/1437/commits/e6bb5d5b957df2cf2b053be52b8359567374cd94). Big diff:
  - For each `header-inl.cuh`, add `header.cuh` and `header-ext.cuh`
  - Remove specialization source files. 
  - Add back source files with instances for each `-ext` header.
  - The only specializations that do not have a corresponding new instance are those in `src/distance/specializations/detail/kernels`. They were in fact not used in the tests, one did not compile, and adding instances would have caused conflicts with PR #1296. This was decided with @tfeher  [here](https://github.com/rapidsai/raft/pull/1415#discussion_r1169302147).
- [Deprecate specialization headers](https://github.com/rapidsai/raft/pull/1437/commits/ff79abf0860321985cee7228bef0afadaf4f3d5d): we do not remove the headers, but just add a `#pragma message`.

**Fixup**: Add some more template instances, remove unnecessary functions, fix a test. 
- [c9e7413](https://github.com/rapidsai/raft/pull/1437/commits/c9e741342b9382a60fe78fc1a9bc11c00c15cb09) Add instances for ivf_flat_interleaved_scan. 
- [Separate fused_l2_nn_helpers](https://github.com/rapidsai/raft/pull/1437/commits/0c889dc64a676287f43db60b884e26d882a2a278): This enables this PR to be **non-breaking**, as compared to PR #1415 . 
- [Remove pairwise_matrix_instantiation_point](https://github.com/rapidsai/raft/pull/1437/commits/f97b2a834a5cdc67894a2488c6850fd9cd09d2df): it is not used anymore. 
- [Rename specialization => instantiation](https://github.com/rapidsai/raft/pull/1437/commits/fb637f7a3cd0d78941f38f3afd67653ff2949b6c): Make usage of `instantiation` in code comments consistent with docs.
- [test/neighbors/selection.cu: Expose kFaissMaxK](https://github.com/rapidsai/raft/pull/1437/commits/7b065aff81a0b1976e2a9e2f3de6690361a1111b) Fix test for `selection.cu`. If it were built with `RAFT_EXPLICIT_INSTANTIATE_ONLY`, then `kFaissMax` would not be defined. By splitting it out into a separate header, it can be used regardless.

### CODE REVIEW

**Comment by ahendriksen** [comment]
Review tip: I have moved the headers by adding the `-inl` suffix in the first commit. The diff is significantly smaller if you start from there.

File: cpp/src/distance/specializations/detail/canberra_double_double_double_int.cu
Comment (cjnolet) [inline cpp/src/distance/specializations/detail/canberra_double_double_double_int.cu:6]: Users downstream are still including these files in their code. Instead of removing all of these files, let's just remove their contents and add a deprecation warning that the files will be removed in 23.08 (or 23.10). The improvements you are making are already pretty massive, but I want to eliminate removing any files that are going to require synchronously updating all of the users before they get merged. In addition, we now have users outside of RAPIDS and our of courtesy we should be giving them time to update their code.

File: cpp/include/raft/neighbors/brute_force-ext.cuh
Comment (ahendriksen) [inline cpp/include/raft/neighbors/brute_force-ext.cuh:83]: Related issue about these instances: https://github.com/rapidsai/raft/issues/1449

**Comment by tfeher** [comment]
> Review tip: I have moved the headers by adding the -inl suffix in the first commit. The diff is significantly smaller if you start from there.

Indeed filtering the commits help tremendously. We have large mechanical changes in, the following commits:
- first commit: [MV: add -inl suffix to header paths](https://github.com/rapidsai/raft/pull/1437/commits/d9801e87f266554c2c561490faf0fe5d76e66165)
- 10th commit: [Replace specializations by split headers](https://github.com/rapidsai/raft/pull/1437/commits/e6bb5d5b957df2cf2b053be52b8359567374cd94)
- 11th commit [Deprecate specialization headers](https://github.com/rapidsai/raft/pull/1437/commits/ff79abf0860321985cee7228bef0afadaf4f3d5d)

It make sense to review these separately. The rest (commits 2-9, 12-16) can be then lumped together during review.

**Review by tfeher — APPROVED** [review]
Thanks Allard for breaking up the oiriginal PR (#1415) into multiple pieces! Also thanks for the detailed description of the changes in the PR description. The questions I had for #1415 are already answered here (or a tracking issue is created). I have one question left, but that could be answered in a follow up PR.

Overall the PR looks good to me, and I am excited about the compile time improvements that we can get once we go through all the steps outlined here and in he follow up PRs.

File: cpp/include/raft/distance/distance-ext.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/distance-ext.cuh:491]: That is a good question. getWorkspaceSize is not an expensive template, so that's not the reason it is instantiated. The problem is that it is defined in the same header as the `distance` function template. To not instantiate it, we can do two things:
1. Move getworkspaceSize to a `distance_helpers.cuh` header and include it in `distance-ext.cuh` and `distance-inl.cuh`. This is what is done for the `fusedL2NN` `initialize` function.
2. Instead of defining `getWorkspaceSize` as `RAFT_EXPLICIT`, copy the definition from the `distance-inl.cuh` header. I think this is a disaster waiting to happen, as it will require a lot of attentiveness to make sure the two definitions do not diverge.

I believe the current concensus is to deprecate getWorkspaceSize anyway, so the long list of instances should be only a temporary issue.

**Review by benfred — COMMENTED** [review]
These changes look great to me - I'm really looking forward to having this series of PR's merged, I think the improvements to build times will make RAFT development much easier.

I'd like to see the corresponding downstream PR's pass CI before getting this merged (like https://github.com/rapidsai/cuml/pull/5363) - do you know why its failing with cuda 11.2/11.4 but working with 11.8? The failure looks like a missing symbol w/ the RusselRao distance 

```python
In [3]: import cpp_demangle

In [4]: cpp_demangle.demangle("_ZN4raft8distance8distanceILNS0_12DistanceTypeE18EfffiEEvRKNS_9resourcesEPKT0_S8_PT2_T3_SB_SB_bS6_")
Out[4]: 'void raft::distance::distance<(raft::distance::DistanceType)18, float, float, float, int>(raft::resources const&, float const*, float const*, float*, int, int, int, bool, float)'
```

**Comment by cjnolet** [comment]
> do you know why its failing with cuda 11.2/11.4 but working with 11.8? 

@vyasr  actually fixed this one today. It turns out some recent changes to the conda packages were causing the cuda toolkit to get pinned to a specific version (11.8) and it was causing cuml's CI to pull in an old version of raft packages for other cuda versions. 

Cuml's CI *should* run fine now.

**Review by benfred — APPROVED** [review]

**Comment by tfeher** [comment]
Some of the headers was split under the `detail` folder, like (like `select_k.cuh`) while others were above the `detail` folder (public level).  @ahendriksen was it done this way to minimize the changes, or was there any other motivation?

The public API did not change in any of the cases, so functionally the current status is fine. But I think it is nicer if we split the files under the detail folder when possible. I would prefer if we do not hold up this PR for such cosmetic changes, but we should consider returning to this question.

**Comment by ahendriksen** [comment]
>Some of the headers was split under the detail folder, like (like select_k.cuh) while others were above the detail folder (public level). @ahendriksen was it done this way to minimize the changes, or was there any other motivation?

Splitting at detail rather than at public level was done for three reasons: 
- The detail header was used internally: for instance the ivf_flat_interleaved_scan_kernel is called in the ivf_flat header as well as the refine header.
- Kind of accidentally: the coalesced reduction kernel was [duplicated many many times](https://github.com/rapidsai/raft/issues/1372) so I added an instance. In hindsight, it might be better to instantiate at a higher level (for instance `linalg::reduce`, `linalg::norm`, etc). To tease apart which `raft::linalg`  kernels are worth instantiating and how could be investigated in the future. The coalesced reduction instances show how it could be done.
- As part of design: the `raft::distance` kernels were optimized for compile time previously. Compiling the kernels without the public header includes can be substantially faster.



> But I think it is nicer if we split the files under the detail folder when possible. 

I think we should favor splitting at the public level. This is beneficial for two reasons: (1) to document which instances exist and (2) to allow downstream libraries to instantiate templates themselves. Instancing at public level allows downstream libraries to opt in to `RAFT_EXPLICIT_INSTANTIATE_ONLY`: when they use a template that has not been instantiated in `libraft.so`, they can instantiate it themselves. Here, it is better that these templates are part of the public API so they don't break as often.

**Comment by tfeher** [comment]
> I think we should favor splitting at the public level. This is beneficial for two reasons: (1) to document which instances exist  ,,,,

Good points!

**Review by divyegala — APPROVED** [review]

File: cpp/include/raft/neighbors/detail/ivf_flat_search-inl.cuh
Comment (cjnolet) [inline cpp/include/raft/neighbors/detail/ivf_flat_search-inl.cuh:1021]: Wow, we need to start breaking headers like this up. Maybe separating out the `kernels` into their own headers.

File: docs/source/developer_guide.md
Comment (cjnolet) [inline docs/source/developer_guide.md:276]:
```suggestion
| defined       |                                | Downstream libraries depending on `libraft` like cuML, cuGraph.                                      |
```

File: docs/source/developer_guide.md
Comment (cjnolet) [inline docs/source/developer_guide.md:344]:
```suggestion
1. In the `-ext.cuh` header, do not include implementation headers. Only include function parameter types and types that are used to instantiate the templates. If a primitive takes custom parameter types, define them in a separate header called `<primitive_name>_types.hpp` (see [Common Design Considerations](https://github.com/rapidsai/raft/blob/7b065aff81a0b1976e2a9e2f3de6690361a1111b/docs/source/developer_guide.md#common-design-considerations)).  
```

File: docs/source/developer_guide.md
Comment (cjnolet) [inline docs/source/developer_guide.md:346]:
```suggestion
2. Keep docstrings in the `-inl.cuh` header, as it is closer to the code. Remove docstrings from template definitions in the `-ext.cuh` header. Make sure to explicitly include public APIs in the RAFT API docs (e.g. that they point the user to the `expensive.cuh` header and not `expensive-inl.cuh`).
```

File: docs/source/using_libraft.md
Comment (cjnolet) [inline docs/source/using_libraft.md:5]:
```suggestion
For most functions, compile-time overhead is minimal but some of RAFT's APIs take a substantial time to compile. As a rule of thumb, most functionality in `raft::distance`, `raft::neighbors`, and `raft::cluster` is expensive to compile and most functionality in other namespaces has little compile-time overhead.
```

Comment (cjnolet) [reply docs/source/using_libraft.md:5]: Please don't reference `raft::spatial` in the docs anymore.

File: docs/source/using_libraft.md
Comment (cjnolet) [inline docs/source/using_libraft.md:21]:
```suggestion
   libraries and use the growing set of `raft::runtime` APIs.
```

Comment (cjnolet) [reply docs/source/using_libraft.md:21]: Can we link to the section on "Runtime APIs" here please?

**Review by cjnolet — CHANGES_REQUESTED** [review]
Really looking forward to these changes, @ahendriksen! I've gone through all the files and I think everything looks good, with the exception of some trivial things (mostly docs related).

File: cpp/include/raft/distance/distance.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/distance.cuh:18]: I misremembered a detail about `ifndef`. I thought it was standardized in C++ 23, but apparently it's `elifdef` and `elifndef` that are standardized in C++ 23. 

https://en.cppreference.com/w/cpp/preprocessor/conditional

File: cpp/include/raft/distance/distance-inl.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/distance-inl.cuh:31]: Good point. I checked this previously, and when I build the docs locally, the docs point to the correct header. For instance: the documentation for `raft::neighbors::ivf_flat::extend` starts with `#include <raft/neighbors/ivf_flat.cuh>`.

I don't know why or how though. I couldn't find any documentation in Doxygen that it automatically points to a non `inl` header.

File: cpp/include/raft/neighbors/detail/ivf_flat_search-inl.cuh
Comment (ahendriksen) [reply cpp/include/raft/neighbors/detail/ivf_flat_search-inl.cuh:1021]: This header is broken up in #1440 :+1:

File: docs/source/developer_guide.md
Comment (ahendriksen) [reply docs/source/developer_guide.md]: Do you happen to know how we can make the link always resolve to our latest branch? Now it is pointing to this specific commit. I could point it to the 23.06 branch, but that will be outdated in a few months too...

File: cpp/include/raft/distance/distance-inl.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/distance-inl.cuh:31]: I now see that we are manually writing the include header in the `.rst` files. So as long as we point users to the correct headers there, this should not be an issue. There seems to be no secret feature in doxygen/sphinx to automatically point to the correct header.

File: docs/source/using_libraft.md
Comment (ahendriksen) [reply docs/source/using_libraft.md]: I couldn't find any documentation of the Runtime APIs. I have filed an [issue](https://github.com/rapidsai/raft/issues/1468) to keep track of it.

File: cpp/include/raft/distance/distance.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/distance.cuh]: I have now replaced `# if !defined` by `#ifndef`.

**Pull request closed without merging**
